File: lib/rbbt/entity/study/plots.rb
input :cutoff, :integer, "Pixels of image", 2
input :size, :integer, "Pixels of image", 14
task :gene_mutation_plot => :binary do |cutoff, size|
  png_file = file(study + ".png")
  FileUtils.mkdir_p File.dirname png_file unless File.exists? File.dirname png_file
  study.R "
library(ggplot2)
library(plyr)
library(reshape)

layer.mutations = rbbt.SE.plot.mutations('#{study}', cutoff=#{cutoff});
p <- ggplot() + layer.mutations 

p <- p + opts(axis.text.x=theme_text(angle=90), panel.background = theme_rect(fill='white', colour='steelblue'))

ggsave(p, filename='#{png_file}', height=#{size}, width=#{size});
"
  Open.read(png_file, :mode => 'rb')
end


input :database, :string, "Database code", :kegg
input :size, :integer, "Pixels of image", 14
task :pathway_mutation_plot => :binary do |database,size|
  png_file = file(study + ".png")
  FileUtils.mkdir_p File.dirname png_file unless File.exists? File.dirname png_file
  study.R "
library(ggplot2)
library(plyr)
library(reshape)


study = '#{study}'
# Sample mutations
sample.mutated.genes = rbbt.SE.sample.mutated.genes(study)
sample.mutated.genes$Sample = rownames(sample.mutated.genes)

# Pathway enrichment
pathway.enrichment = rbbt.ruby.substitutions(
    \"
    require 'rbbt/workflow'
    require 'rbbt/entity'
    require 'rbbt/entity/gene'
    require 'rbbt/sources/pfam'
    require 'rbbt/sources/kegg'
    require 'rbbt/sources/go'

    YAML::ENGINE.yamler = 'syck' if defined? YAML::ENGINE and YAML::ENGINE.respond_to? :yamler

    Workflow.require_workflow 'StudyExplorer'

    study = Study.setup('STUDY')

    pathways = study.job(:mutation_pathway_enrichment, study, :baseline => :pathway_base_counts, :database => '#{database}', :fdr => false).run.select('p-value'){|pvalue| pvalue = pvalue.first.to_f if Array === pvalue; pvalue < 0.2}
    pathways.add_field 'Name' do |pathway, values|
        [pathway.name]
    end

    pathways.add_field 'Gene' do |pathway, values|
        values['Ensembl Gene ID'].name
    end

    pathways = pathways.select('Name'){|name| name.first.to_s !~ /cancer|olfactory|glioma|melanoma|malaria|leukemia|carcinoma|sarcoma/i}

    \", substitutions=list(STUDY=study));


# Sample pathway mutations
find.mutated.pathways.for.sample <- function(x, pathway.info){
    all.genes = names(x);
    genes = all.genes[x==TRUE];
    ddply(pathway.info, 'Name', function(x){pathway.genes = unlist(strsplit(x$Gene, '\\\\|')); if (length(intersect(genes, pathway.genes)) > 0){TRUE}else{FALSE}})
}
sample.pathway.mutations = ddply(sample.mutated.genes, 'Sample', find.mutated.pathways.for.sample, pathway.info = pathway.enrichment)
names(sample.pathway.mutations) = c('Sample', 'Pathway', 'Mutated')

p <- ggplot(sample.pathway.mutations) + geom_tile(aes(x=Sample, y=Pathway, alpha=Mutated))

p <- rbbt.SE.plot.sort.by.pathway.mutations(p)


# Mark repeated genes


d = p$data
d$Exclusive = FALSE

pathway.genes = list();
for(pathway in levels(d$Pathway)){
   pathway.genes[pathway] = strsplit(pathway.enrichment[pathway.enrichment[,'Name'] == pathway, 'Gene'], '\\\\|')
}

find.exclusive.pathway.genes <- function(data, pathways){
  found.genes = c();
  exclusive.pathway.genes = list();
  sample = as.character(unique(data$Sample));
  for(pathway in pathways){
     current.pathway.genes = pathway.genes[[pathway]];
     sample.genes = names(sample.mutated.genes)[sample.mutated.genes[sample,] == TRUE]
     sample.pathway.genes = intersect(current.pathway.genes, sample.genes);
     exclusive.genes = setdiff(sample.pathway.genes, found.genes);
     found.genes = c(found.genes, exclusive.genes)
     exclusive.pathway.genes[[pathway]] = exclusive.genes
  }

  return(exclusive.pathway.genes);
}

exclusive.pathway.genes = dlply(d, 'Sample', find.exclusive.pathway.genes, pathways = levels(d$Pathway)) 

for( sample in names(exclusive.pathway.genes)){
     pathway.exclusive.genes = exclusive.pathway.genes[[sample]];
     for( pathway in names(pathway.exclusive.genes)){
        if (length(pathway.exclusive.genes[[pathway]]) > 0){
           print(sample)
           print(pathway)
           d[(d$Sample == sample & d$Pathway == pathway), 'Exclusive'] = TRUE
        }
     }
}

p$data = d


p <- p + aes(fill=Exclusive)

p <- p + opts(axis.text.x=theme_text(angle=90), panel.background = theme_rect(fill='white', colour='steelblue'))

p




ggsave(p, filename='#{png_file}', height=#{size}, width=#{size});
"
  Open.read(png_file, :mode => 'rb')
end

File: share/R/plots.R
rbbt.SE.plot.sort.by.field <- function(plot, field){
    d = plot$data;

    d[[field]] = reorder(d[[field]], d$Mutated, sum)

    sample.best.gene.pos.df = ddply(d, "Sample", function(x){ min(match(subset(x, Mutated==TRUE)[[field]], rev(levels(d[[field]]))), na.rm=T)})

    d$sample.best.gene.pos = NULL
    names(sample.best.gene.pos.df) <- c("Sample", "sample.best.gene.pos");

    d = merge(d, sample.best.gene.pos.df, all.x=TRUE)

    d$Sample = reorder(d$Sample, d$sample.best.gene.pos)

    plot$data = d;

    return(plot);
}

rbbt.SE.plot.sort.by.mutations <- function(plot){
    d = plot$data;

    d$Gene = reorder(d$Gene, d$Mutated, sum);
    num.elems = length(levels(d$Gene));

    #sample.best.gene.pos.df = ddply(d, "Sample", function(x){ 1/mean(1/match(subset(x, Mutated==TRUE)$Gene, rev(levels(d$Gene)))^2)})
    sample.best.gene.pos.df = ddply(d, "Sample", function(x){ 1/sum(2^(num.elems - match(subset(x, Mutated==TRUE)$Gene, rev(levels(d$Gene)))))})

    d$sample.best.gene.pos = NULL
    names(sample.best.gene.pos.df) <- c("Sample", "sample.best.gene.pos");

    d = merge(d, sample.best.gene.pos.df, all.x=TRUE)

    d$Sample = reorder(d$Sample, d$sample.best.gene.pos)

    plot$data = d;

    return(plot);
}

rbbt.SE.plot.sort.by.pathway.mutations <- function(plot){
    d = plot$data;

    d$Pathway = reorder(d$Pathway, d$Mutated, sum);
    num.elems = length(levels(d$Pathway));

    #sample.best.gene.pos.df = ddply(d, "Sample", function(x){ min(match(subset(x, Mutated==TRUE)$Pathway, rev(levels(d$Pathway))), na.rm=T)})
    #sample.best.gene.pos.df = ddply(d, "Sample", function(x){ 1/mean(1/match(subset(x, Mutated==TRUE)$Pathway, rev(levels(d$Pathway)))^2)})
    sample.best.gene.pos.df = ddply(d, "Sample", function(x){ 1/sum(2^(num.elems - match(subset(x, Mutated==TRUE)$Pathway, rev(levels(d$Pathway)))))})

    d$sample.best.gene.pos = NULL
    names(sample.best.gene.pos.df) <- c("Sample", "sample.best.gene.pos");

    d = merge(d, sample.best.gene.pos.df, all.x=TRUE)

    d$Sample = reorder(d$Sample, d$sample.best.gene.pos)

    plot$data = d;

    return(plot);
}

rbbt.SE.plot.mutations <- function(study, cutoff = 3, sample.info = NULL){
    sample.mutated.genes = rbbt.SE.sample.mutated.genes(study);

    gene.mutation.counts = apply(sample.mutated.genes, 2, function(x){sum(x==TRUE)})

    recurrent.genes = names(gene.mutation.counts[gene.mutation.counts >= cutoff])

    d.recurrent = sample.mutated.genes[, recurrent.genes]
    d.recurrent$Sample = rownames(d.recurrent)

    d.recurrent.m = melt(d.recurrent, "Sample")

    names(d.recurrent.m) <- c("Sample", "Gene", "Mutated")

    if (is.null(sample.info)){
        d = d.recurrent.m
    }else{
        d = merge(d.recurrent.m, sample.info, all.x=TRUE)
    }

    layer.mutations = geom_tile(data=d,aes(x=Sample, y=Gene, alpha=Mutated))

    rbbt.SE.plot.sort.by.mutations(layer.mutations);

    return(layer.mutations);
}

rbbt.SE.plot.add.expression <- function(plot, study, ...){

    genes = unique(plot$data$Gene);
    gene.expression <- rbbt.SE.expression(study, genes, ...);

    gene.expression.m <- melt(gene.expression);
    names(gene.expression.m) <- c("Gene", "Sample", "Expression");

    gene.expression.3rd = summary(gene.expression.m$Expression)[["3rd Qu."]]
    gene.expression.1st = summary(gene.expression.m$Expression)[["1st Qu."]]

    unpadd = as.character(as.numeric(gene.expression.m$Sample));
    unpadd[is.na(unpadd)] = gene.expression.m$Sample[is.na(unpadd)];
    gene.expression.m$Sample = unpadd;

    mean.gene.expression <- aggregate(Expression ~ Gene, gene.expression.m, mean, trim=0.1, na.rm=T);
    names(mean.gene.expression) <- c("Gene", "Mean");
    gene.expression.m[gene.expression.m[,"Expression"] > gene.expression.3rd, "Expression"] = gene.expression.3rd
    gene.expression.m[gene.expression.m[,"Expression"] < gene.expression.1st, "Expression"] = gene.expression.1st

    sd.gene.expression <- aggregate(Expression ~ Gene, gene.expression.m, mad, na.rm=T);
    names(sd.gene.expression) <- c("Gene", "SD");

    gene.expression.m <- merge(gene.expression.m, mean.gene.expression);
    gene.expression.m <- merge(gene.expression.m, sd.gene.expression);

    #d = merge(d, gene.expression.m, by=c("Sample", "Gene"), all.x=TRUE);
    plot$data = merge(plot$data, gene.expression.m, all.x=TRUE);

    layer.expression = geom_point(data=plot$data, aes(x=Sample, y=Gene, size=abs((Expression - Mean) / SD), color=((Expression - Mean) / SD)));
 
    return(layer.expression)
}
